File: skills/singlecell-qc/scripts/calculate_metrics.R
#!/usr/bin/env Rscript
# Per-sample scRNA QC metric calculation (Matrix / base R).
#
# Human-in-the-loop tool: computes metrics and writes metadata.tsv.
# Does NOT apply filtering thresholds.
#
# Bundled with singlecell-qc skill.
# Example:
#   Rscript calculate_metrics.R \
#     --matrix-dir data/cellranger/SAMPLE/outs/raw_feature_bc_matrix \
#     --sample-id SAMPLE \
#     --species human \
#     --output-dir result/01-qc/01-calculate_metrics/SAMPLE

suppressPackageStartupMessages({
  if (!requireNamespace("jsonlite", quietly = TRUE)) {
    stop("Package 'jsonlite' is required.")
  }
  if (!requireNamespace("Matrix", quietly = TRUE)) {
    stop("Package 'Matrix' is required.")
  }
})

`%||%` <- function(a, b) if (!is.null(a)) a else b

script_dir <- {
  args <- commandArgs(trailingOnly = FALSE)
  file_arg <- sub("^--file=", "", args[grep("^--file=", args)])
  if (length(file_arg) > 0) {
    dirname(normalizePath(file_arg[[1]], winslash = "/", mustWork = FALSE))
  } else {
    getwd()
  }
}
skill_root <- normalizePath(file.path(script_dir, ".."), winslash = "/", mustWork = FALSE)
if (!dir.exists(file.path(skill_root, "assets", "gene_sets"))) {
  configured_root <- Sys.getenv("SINGLECELL_QC_SKILL_ROOT", unset = "")
  if (nzchar(configured_root)) {
    skill_root <- normalizePath(configured_root, winslash = "/", mustWork = FALSE)
  }
}
if (!dir.exists(file.path(skill_root, "assets", "gene_sets"))) {
  stop("Cannot locate bundled gene sets. Set SINGLECELL_QC_SKILL_ROOT to the installed skill directory.")
}
gene_sets <- file.path(skill_root, "assets", "gene_sets")

parse_args <- function() {
  args <- commandArgs(trailingOnly = TRUE)
  out <- list(
    matrix_dir = NULL,
    sample_id = NULL,
    species = "human",
    output_dir = NULL,
    min_genes = 50,
    hq_n_genes = 500
  )
  i <- 1
  while (i <= length(args)) {
    key <- args[[i]]
    if (key %in% c("--matrix-dir", "--sample-id", "--species", "--output-dir")) {
      out[[gsub("-", "_", sub("^--", "", key))]] <- args[[i + 1]]
      i <- i + 2
    } else if (key == "--min-genes") {
      out$min_genes <- as.integer(args[[i + 1]]); i <- i + 2
    } else if (key == "--hq-n-genes") {
      out$hq_n_genes <- as.integer(args[[i + 1]]); i <- i + 2
    } else if (key %in% c("-h", "--help")) {
      cat(paste(
        "Usage: Rscript calculate_metrics.R",
        "--matrix-dir DIR --sample-id ID --output-dir DIR",
        "[--species human|mouse] [--min-genes 50] [--hq-n-genes 500]\n"
      ))
      quit(save = "no", status = 0)
    } else {
      stop("Unknown argument: ", key)
    }
  }
  if (is.null(out$matrix_dir) || is.null(out$sample_id) || is.null(out$output_dir)) {
    stop("Required: --matrix-dir, --sample-id, --output-dir")
  }
  out
}

read_gene_list <- function(path) {
  if (!file.exists(path)) return(character())
  lines <- readLines(path, warn = FALSE)
  lines <- trimws(lines)
  lines[nzchar(lines) & !grepl("^#", lines)]
}

read_10x_counts <- function(matrix_dir) {
  counts <- Matrix::readMM(file.path(matrix_dir, "matrix.mtx.gz"))
  counts <- as(counts, "CsparseMatrix")
  features <- read.table(file.path(matrix_dir, "features.tsv.gz"), header = FALSE)
  barcodes <- read.table(file.path(matrix_dir, "barcodes.tsv.gz"), header = FALSE)
  rownames(counts) <- make.unique(features$V2)
  colnames(counts) <- barcodes$V1
  counts
}

fraction_score <- function(counts, genes, denom) {
  genes <- intersect(genes, rownames(counts))
  if (length(genes) == 0) return(rep(0, ncol(counts)))
  num <- Matrix::colSums(counts[genes, , drop = FALSE])
  out <- num / denom
  out[!is.finite(out)] <- 0
  out
}

opts <- parse_args()
counts <- read_10x_counts(opts$matrix_dir)

n_genes <- Matrix::colSums(counts > 0)
keep <- n_genes >= opts$min_genes
counts <- counts[, keep, drop = FALSE]
n_genes <- n_genes[keep]

metadata <- data.frame(
  sample_id = opts$sample_id,
  barcode = colnames(counts),
  n_genes = as.integer(n_genes),
  n_UMIs = as.numeric(Matrix::colSums(counts)),
  stringsAsFactors = FALSE,
  row.names = colnames(counts)
)

mt_pattern <- if (identical(opts$species, "mouse")) "^mt-" else "^MT-"
mito_genes <- grep(mt_pattern, rownames(counts), value = TRUE)
rb_genes <- grep("^(RPS|RPL)", rownames(counts), value = TRUE)

metadata$mito_frac <- fraction_score(counts, mito_genes, metadata$n_UMIs)
metadata$pct_counts_rb <- 100 * fraction_score(counts, rb_genes, metadata$n_UMIs)

hbb_file <- file.path(gene_sets, paste0("hbb_genes_", opts$species, ".txt"))
metadata$hbb_score <- fraction_score(counts, read_gene_list(hbb_file), metadata$n_UMIs)

if (identical(opts$species, "human")) {
  chry_file <- file.path(gene_sets, "chrY_genes_human.txt")
  metadata$chrY_frac <- fraction_score(counts, read_gene_list(chry_file), metadata$n_UMIs)
}

metadata$is_HQ <- metadata$n_genes >= opts$hq_n_genes
metadata$ambient_frac <- NA_real_
metadata$doublet_score <- NA_real_
metadata$predicted_doublet <- FALSE
metadata$phase <- NA_character_
metadata$s_score <- NA_real_
metadata$g2m_score <- NA_real_

# Optional extensions — enable in project scripts when packages/data exist:
# - celda::decontX for ambient_frac (HQ cells)
# - DoubletFinder / scDblFinder for doublet_score
# - Seurat::CellCycleScoring for phase/s_score/g2m_score
# - STARsolo velocyto for nuclear_frac

dir.create(opts$output_dir, recursive = TRUE, showWarnings = FALSE)
write.table(
  metadata,
  file = file.path(opts$output_dir, "metadata.tsv"),
  sep = "\t", quote = FALSE, row.names = TRUE, col.names = TRUE
)

summary <- list(
  sample_id = opts$sample_id,
  n_cells = nrow(metadata),
  median_n_genes = stats::median(metadata$n_genes),
  median_n_UMIs = stats::median(metadata$n_UMIs),
  median_mito_frac = stats::median(metadata$mito_frac),
  n_HQ = sum(metadata$is_HQ)
)
jsonlite::write_json(summary, file.path(opts$output_dir, "metrics_summary.json"), auto_unbox = TRUE, pretty = TRUE)

message("Wrote metrics for ", opts$sample_id, ": ", nrow(metadata), " cells -> ", opts$output_dir)

File: skills/singlecell-qc/scripts/calculate_metrics.py
#!/usr/bin/env python3
"""Per-sample scRNA QC metric calculation (scanpy).

Human-in-the-loop tool: computes metrics and writes metadata.tsv.
Does NOT apply filtering thresholds. Run filter scripts only after
the analyst reviews diagnosis outputs and confirms cutoffs.

Bundled with singlecell-qc skill.

Example:
  python calculate_metrics.py \\
    --matrix-dir data/cellranger/SAMPLE/outs/raw_feature_bc_matrix \\
    --sample-id SAMPLE \\
    --species human \\
    --output-dir result/01-qc/01-calculate_metrics/SAMPLE
"""

from __future__ import annotations

import argparse
import gzip
import json
from pathlib import Path

import numpy as np
import pandas as pd
import scanpy as sc
from scipy import sparse


SKILL_ROOT = Path(__file__).resolve().parents[1]
GENE_SETS = SKILL_ROOT / "assets" / "gene_sets"


def read_gene_list(path: Path) -> list[str]:
    genes = []
    for line in path.read_text().splitlines():
        g = line.strip()
        if g and not g.startswith("#"):
            genes.append(g)
    return genes


def fraction_score(adata, genes: list[str], denom: np.ndarray) -> np.ndarray:
    if not genes:
        return np.zeros(adata.n_obs, dtype=float)
    sub = adata[:, genes].X
    if sparse.issparse(sub):
        num = np.asarray(sub.sum(axis=1)).ravel()
    else:
        num = sub.sum(axis=1)
    with np.errstate(divide="ignore", invalid="ignore"):
        out = num / denom
    out[~np.isfinite(out)] = 0.0
    return out


def load_10x(matrix_dir: Path, min_genes: int = 50) -> sc.AnnData:
    adata = sc.read_10x_mtx(
        str(matrix_dir),
        var_names="gene_symbols",
        make_unique=True,
    )
    if min_genes > 0:
        sc.pp.filter_cells(adata, min_genes=min_genes)
    return adata


def compute_metrics(
    adata: sc.AnnData,
    species: str,
    hq_n_genes: int,
    run_scrublet: bool,
    expected_doublet_rate: float,
) -> pd.DataFrame:
    mt_prefix = "MT-" if species == "human" else "mt-"
    adata.var["mt"] = adata.var_names.str.startswith(mt_prefix)
    adata.var["rb"] = adata.var_names.str.startswith(("RPS", "RPL"))

    sc.pp.calculate_qc_metrics(
        adata,
        qc_vars=["mt", "rb"],
        percent_top=None,
        log1p=False,
        inplace=True,
    )

    obs = adata.obs.copy()
    obs["n_genes"] = obs["n_genes_by_counts"].astype(int)
    obs["n_UMIs"] = obs["total_counts"].astype(float)
    obs["mito_frac"] = obs["pct_counts_mt"].astype(float) / 100.0
    obs["pct_counts_rb"] = obs["pct_counts_rb"].astype(float)
    obs["is_HQ"] = obs["n_genes"] >= hq_n_genes

    hbb_file = GENE_SETS / f"hbb_genes_{species}.txt"
    if hbb_file.exists():
        hbb = [g for g in read_gene_list(hbb_file) if g in adata.var_names]
        obs["hbb_score"] = fraction_score(adata, hbb, obs["n_UMIs"].to_numpy())

    if species == "human":
        chry_file = GENE_SETS / "chrY_genes_human.txt"
        if chry_file.exists():
            chry = [g for g in read_gene_list(chry_file) if g in adata.var_names]
            obs["chrY_frac"] = fraction_score(adata, chry, obs["n_UMIs"].to_numpy())

    obs["doublet_score"] = np.nan
    obs["predicted_doublet"] = False

    if run_scrublet and obs["is_HQ"].sum() >= 100:
        import scrublet as scr

        hq_idx = obs.index[obs["is_HQ"]]
        hq = adata[hq_idx]
        x = hq.X.toarray() if hasattr(hq.X, "toarray") else np.asarray(hq.X)
        scrub = scr.Scrublet(x, expected_doublet_rate=expected_doublet_rate)
        scores, preds = scrub.scrub_doublets(
            min_counts=2,
            min_cells=3,
            min_gene_variability_pctl=85,
            n_prin_comps=min(30, hq.n_obs - 1, hq.n_vars - 1),
        )
        obs.loc[hq_idx, "doublet_score"] = scores
        obs.loc[hq_idx, "predicted_doublet"] = preds

    return obs


def write_metadata(obs: pd.DataFrame, out_dir: Path, sample_id: str) -> None:
    out_dir.mkdir(parents=True, exist_ok=True)
    obs = obs.copy()
    obs.insert(0, "barcode", obs.index.astype(str))
    obs.insert(0, "sample_id", sample_id)

    tsv = out_dir / "metadata.tsv"
    obs.to_csv(tsv, sep="\t")

    with gzip.open(out_dir / "metadata.tsv.gz", "wt") as fh:
        obs.to_csv(fh, sep="\t")

    summary = {
        "sample_id": sample_id,
        "n_cells": int(obs.shape[0]),
        "median_n_genes": float(np.median(obs["n_genes"])),
        "median_n_UMIs": float(np.median(obs["n_UMIs"])),
        "median_mito_frac": float(np.median(obs["mito_frac"])),
        "n_HQ": int(obs["is_HQ"].sum()),
        "n_predicted_doublet": int(obs["predicted_doublet"].sum()),
    }
    (out_dir / "metrics_summary.json").write_text(json.dumps(summary, indent=2))


def parse_args() -> argparse.Namespace:
    p = argparse.ArgumentParser(description="Per-sample scRNA QC metrics (scanpy)")
    p.add_argument("--matrix-dir", required=True, type=Path, help="10x matrix directory")
    p.add_argument("--sample-id", required=True, help="Sample identifier")
    p.add_argument("--species", choices=["human", "mouse"], default="human")
    p.add_argument("--output-dir", required=True, type=Path)
    p.add_argument("--min-genes", type=int, default=50, help="Pre-filter min genes")
    p.add_argument("--hq-n-genes", type=int, default=500, help="HQ gate for Scrublet")
    p.add_argument("--run-scrublet", action="store_true", help="Run Scrublet on HQ cells")
    p.add_argument("--expected-doublet-rate", type=float, default=0.06)
    p.add_argument("--save-h5ad", action="store_true", help="Save AnnData with obs metrics")
    return p.parse_args()


def main() -> None:
    args = parse_args()
    adata = load_10x(args.matrix_dir, min_genes=args.min_genes)
    obs = compute_metrics(
        adata,
        species=args.species,
        hq_n_genes=args.hq_n_genes,
        run_scrublet=args.run_scrublet,
        expected_doublet_rate=args.expected_doublet_rate,
    )
    adata.obs = obs
    write_metadata(obs, args.output_dir, args.sample_id)
    if args.save_h5ad:
        adata.write_h5ad(args.output_dir / "with_metrics.h5ad", compression="gzip")
    print(f"Wrote metrics for {args.sample_id}: {obs.shape[0]} cells -> {args.output_dir}")


if __name__ == "__main__":
    main()
